File: MesiboApplication/SampleAppFileTransferHandler.m
//
//  Mesibo_FileTransferHandler.m
//  LiveChat
//
//  Created by Mesibo on 22/09/17.
//  Copyright © 2018 Mesibo. All rights reserved.
//

#import "SampleAppFileTransferHandler.h"
#import "SampleAPI.h"
#import "MesiboMessenger-Swift.h"

#import <Photos/Photos.h>

@implementation SampleAppFileTransferHandler

- (void) initialize {
    [MesiboInstance addListener:self];
}

-(BOOL) Mesibo_onStartUpload:(MesiboFileInfo *)file {
    int type = [MesiboInstance getNetworkConnectivity];
    
    if(MESIBO_CONNECTIVITY_WIFI != type && !file.userInteraction)
        return NO;
    
    //TBD, check max file size
    MesiboParams *params = [file getParams];
    
    NSMutableDictionary *post = [[NSMutableDictionary alloc] init];
    
    [post setObject:@"upload" forKey:@"op"];
    [post setValue:[SampleAPIInstance getToken] forKey:@"token"];
    [post setValue:[@(params.mid) stringValue] forKey:@"mid"];
    [post setValue:[@(0) stringValue] forKey:@"profile"];
    
    //UIImage *a = [MesiboInstance loadImage:nil filePath:[file getPath] maxside:0];
    
    Mesibo_onHTTPProgress handler = ^BOOL(MesiboHttp *http, int state, int progress) {
        
        int status = [file getStatus];
        
        if(MESIBO_FILESTATUS_RETRYLATER != status) {
            status = MESIBO_FILESTATUS_INPROGRESS;
            if(progress < 0)
                status = MESIBO_FILESTATUS_FAILED;
        }
        
        if(100 == progress && MESIBO_HTTPSTATE_DOWNLOAD == state) {
            NSError *jsonerror = nil;
            
            NSData *data = [[http getDataString] dataUsingEncoding:NSUTF8StringEncoding];
            id jsonObject = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&jsonerror];
            NSDictionary *returnedDict = (NSMutableDictionary *)jsonObject;
            NSString *fileUrl =  [returnedDict valueForKey:@"file"];
            
            if(fileUrl)
                [file setUrl:fileUrl];
            else {
                progress = -1;
                status = MESIBO_FILESTATUS_FAILED;
            }
        }
        
        if(progress < 100 || (100 == progress && MESIBO_HTTPSTATE_DOWNLOAD == state))
           [MesiboInstance updateFileTransferProgress:file progress:progress status:status];
        
        return ((100 == progress && MESIBO_HTTPSTATE_DOWNLOAD == state) || MESIBO_FILESTATUS_RETRYLATER != status);
    };
    
    MesiboHttp *http = [MesiboHttp new];
    http.url = [SampleAPIInstance getUploadUrl];
    http.uploadPhAsset = file.asset;
    http.uploadLocalIdentifier = file.localIdentifier;
    http.uploadFile = [file getPath];
    http.postBundle = post;
    http.uploadFileField = @"photo";
    http.listener =  handler;
    
    [file setFileTransferContext:http];
 
    return [http execute];
}

-(BOOL) Mesibo_onStartDownload:(MesiboFileInfo *)file {
    int type = [MesiboInstance getNetworkConnectivity];
    
    if(![SampleAPIInstance getMediaAutoDownload] && MESIBO_CONNECTIVITY_WIFI != type && !file.userInteraction)
        return NO;
    
    MesiboParams *params = [file getParams];
    
    if(MESIBO_ORIGIN_REALTIME != params.origin && !file.userInteraction)
        return NO;
    
    NSString *url = [file getUrl];
    
    if (![url hasPrefix:@"http://"] && ![url hasPrefix:@"https://"]) {
        url = [[SampleAPIInstance getDownloadUrl] stringByAppendingString:url];
    }
    
    Mesibo_onHTTPProgress handler = ^BOOL(MesiboHttp *http, int state, int progress) {
        int status = [file getStatus];
        
        if(MESIBO_FILESTATUS_RETRYLATER != status) {
            status = MESIBO_FILESTATUS_INPROGRESS;
            if(progress < 0)
                status = MESIBO_FILESTATUS_FAILED;
        }
        
        [MesiboInstance updateFileTransferProgress:file progress:progress status:status];
        return (100 == progress || MESIBO_FILESTATUS_RETRYLATER != status);
    };
    
    MesiboHttp *http = [MesiboHttp new];
    http.url = url;
    http.downloadFile = [file getPath];
    http.resume = YES;
    http.listener =  handler;
    
    [file setFileTransferContext:http];
    
    return [http execute];
    
}

-(BOOL) Mesibo_onStartFileTransfer:(MesiboFileInfo *)file {
    
    if(MESIBO_FILEMODE_DOWNLOAD == file.mode) {
        return [self Mesibo_onStartDownload:file];
    }
    
    return [self Mesibo_onStartUpload:file];
    
}

-(BOOL) Mesibo_onStopFileTransfer:(MesiboFileInfo *)file {
    MesiboHttp *http = [file getFileTransferContext];
    if(http) {
        [http cancel];
    }
    return YES;
}

@end

File: MesiboApplication/SampleAppFileTransferHandler.swift
/******************************************************************************
* By accessing or copying this work, you agree to comply with the following   *
* terms:                                                                      *
*                                                                             *
* Copyright (c) 2019-2023 mesibo                                              *
* https://mesibo.com                                                          *
* All rights reserved.                                                        *
*                                                                             *
* Redistribution is not permitted. Use of this software is subject to the     *
* conditions specified at https://mesibo.com . When using the source code,    *
* maintain the copyright notice, conditions, disclaimer, and  links to mesibo * 
* website, documentation and the source code repository.                      *
*                                                                             *
* Do not use the name of mesibo or its contributors to endorse products from  *
* this software without prior written permission.                             *
*                                                                             *
* This software is provided "as is" without warranties. mesibo and its        *
* contributors are not liable for any damages arising from its use.           *
*                                                                             *
* Documentation: https://mesibo.com/documentation/                            *
*                                                                             *
* Source Code Repository: https://github.com/mesibo/                          *
*******************************************************************************/

import Foundation
import Photos

@objcMembers public class SampleAppFileTransferHandler: NSObject, MesiboDelegate {
    
    func initialize() {
        Mesibo.getInstance().addListener(self)
    }
    
    func mesibo_(onStartUpload file: MesiboFileTransfer) -> Bool {
               
        var post: [AnyHashable : Any] = [:]
        
        post["op"] = "upload"
        post["auth"] = SampleAPI.getInstance().getToken()
        post["mid"] = String(file.mid)
        post["profile"] = "0"
        
        let handler = { (http: MesiboHttp, state: Int32, httpprogress: Int32) -> Bool in
            
            var progress: Int32 = httpprogress
                       
            if 100 == progress && MESIBO_HTTPSTATE_DOWNLOAD == state {
                var jsonerror: Error? = nil
                
                let data = http.getDataString()?.data(using: .utf8)
                var jsonObject: Any? = nil
                do {
                    if let data = data {
                        jsonObject = try JSONSerialization.jsonObject(with: data, options: .allowFragments)
                    }
                } catch let jsonerror {
                }
                let returnedDict = jsonObject as? [AnyHashable : Any]
                let fileUrl = returnedDict?["url"] as? String
                
                if fileUrl != nil {
                    file.setResult(true, url: fileUrl)
                } else {
                    file.setResult(false, url: nil)
                }
                return true;
            }
            
            if (progress < 0 ) {
                file.setResult(false, url: nil)
                return true;
            }
            
            file.progress = progress;
            return true;
            
            } as Mesibo_onHTTPProgress
        
        let http = MesiboHttp()
        http.url = SampleAPI.getInstance().getUploadUrl()
        http.uploadPhAsset = file.getPHAsset()
        http.uploadLocalIdentifier = file.getLocalIdentifier()
        http.uploadFile = file.getPath()
        http.postBundle = post
        http.uploadFileField = "photo"
        http.listener = handler
        
        file.setFileTransferContext(obj: http);
        
        return http.execute()
        
       
    }
    
    func mesibo_(onStartDownload file: MesiboFileTransfer) -> Bool {
   
        if MESIBO_ORIGIN_REALTIME != file.origin && file.priority == 0 {
            return false
        }
        
        var url = file.getUrl()
        
        let handler = { (http: MesiboHttp, state: Int32, progress: Int32) -> Bool in
            
            if(100 == progress) {
                file.setResult(false, url: nil)
                return true;
            }
            
            if (progress < 0 ) {
                file.setResult(false, url: nil)
                return true;
            }
            
            file.progress = progress;
            return true;
            
            } as Mesibo_onHTTPProgress
        
        let http = MesiboHttp()
        http.url = url
        http.downloadFile = file.getPath()
        http.resume = true
        http.listener = handler
        
        file.setFileTransferContext(obj: http);
        
        return http.execute()
        
    }
    
    public func Mesibo_onStartFileTransfer(ft: MesiboFileTransfer) -> Bool {
        
        if (!ft.upload) {
            return mesibo_(onStartDownload: ft)
        }
        
        return mesibo_(onStartUpload: ft)
        
    }
    
    public func Mesibo_onStopFileTransfer(ft: MesiboFileTransfer) -> Bool {
        let http = ft.getFileTransferContext() as? MesiboHttp
        if http != nil {
            http!.cancel()
        }
        return true
    }
}
